### ISSUE TITLE
T2 Unlimited

### ISSUE THREAD

**Issue by bjeanes** [buildkite/elastic-ci-stack-for-aws#437]
Any chance we could expose an option to enable [T2 Unlimited](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/t2-unlimited.html)? In particular I want to enable this automatically on `builder` queues so that a stale cache will let me burst up and use a lot of CPU. I had a lot of failed builds recently and my CPU credits got depleted, blowing out all builds horrendously. I'd love the option to enable T2 Unlimited automatically so I'm not CPU constrained, as over a 24 hour period I'm not likely to constantly be using those credits.

**Comment by lox** [comment]
That seems reasonable to me, happy to accept a PR!

**Comment by bjeanes** [comment]
I'd be willing but I don't know anything about CloudFormation and can't make much sense of the (basically Turing-complete) YAML files that make up this repository 😬... Thing you can give me any tactical pointers?

**Comment by dominics** [comment]
`AWS::AutoScaling::LaunchConfiguration` in CloudFormation apparently has no way to specify a credit specification; you either have to use `AWS::EC2::LaunchTemplate`s or use the `modify-instance-credit-specification` API after the instance is launched.

So, this is harder than it appears at first.

**Comment by lox** [comment]
Any `AWS::EC2::LaunchTemplate` has no way to specify spot pricing 😭

**Comment by pda** [comment]
Without changing the stack template, you could point `BootstrapScriptUrl` at a script that changes the instance credit specifications upon boot, similar to this example:

```sh
https://github.com/suzryo/aws/blob/1166a62d4ff79079a877e0be1bc9648c5501d393/cfn-ec2-t2-unlimited/autoscale-t2-unlimited.yaml#L89-L99
```

I guess for the stack to expose a parameter for this, it could be templated into the `AWS::AutoScaling::LaunchConfiguration` `UserData`.

**Comment by lox** [comment]
Is this basically solved by the `t3` class of instances?

**Comment by bjeanes** [comment]
Yes. I meant to update this actually but my stack kept failing to update to
T3 right after announcement so I forgot about it. In theory, that should
totally address the issue. For the unfamiliar, T3s default to Unlimited
mode.
